core/src/test/scala/caliban/wrappers/WrappersSpec.scala

```diff
@@ -0,0 +1,81 @@
+import caliban.RootResolver
+import caliban.schema.GenericSchema
+import zio.clock.Clock
+import zio.duration._
+import zio.test.Assertion._
+import zio.test._
+import zio.test.environment.TestClock
+import zio.{ clock, URIO }
+
+object WrappersSpec
+    extends DefaultRunnableSpec(
+      suite("WrappersSpec")(
+        testM("Timeout") {
+          case class Test(a: URIO[Clock, Int])
+
+          object schema extends GenericSchema[Clock]
+          import schema._
+
+          val interpreter =
+            Wrappers
+              .timeout(1 minute)(
+                graphQL(RootResolver(Test(clock.sleep(2 minutes).as(0))))
+              )
+              .interpreter
+          val query = gqldoc("""
+              {
+                a
+              }""")
+          assertM(
+            TestClock.adjust(1 minute) *> interpreter.execute(query).map(_.errors),
+            equalTo(List(ExecutionError("""Query was interrupted after timeout of 1 m:
+
+              {
+                a
+              }""".stripMargin)))
+          )
+        },
+        testM("Apollo Tracing") {
+          case class Query(hero: Hero)
+          case class Hero(name: String, friends: List[Hero] = Nil)
+
+          object schema extends GenericSchema[Clock]
+          import schema._
+
+          val interpreter =
+            ApolloTracing
+              .apolloTracing(
+                graphQL(
+                  RootResolver(
+                    Query(Hero("R2-D2", List(Hero("Luke Skywalker"), Hero("Han Solo"), Hero("Leia Organa"))))
+                  )
+                )
+              )
+              .map(_.interpreter)
+
+          val query = gqldoc("""
+              {
+                hero {
+                  name
+                  friends {
+                    name
+                  }
+                }
+              }""")
+          assertM(
+            interpreter >>= (_.execute(query).map(_.extensions.map(_.toString))),
+            isSome(
+              equalTo(
+                """{"tracing":{"version":1,"startTime":"1970-01-01T00:00:00.000Z","endTime":"1970-01-01T00:00:00.000Z","duration":0,"parsing":{"startOffset":0,"duration":0},"validation":{"startOffset":0,"duration":0},"execution":{"resolvers":[{"path":["hero"],"parentType":"Query","fieldName":"hero","returnType":"Hero!","startOffset":0,"duration":0},{"path":["hero","friends"],"parentType":"Hero","fieldName":"friends","returnType":"[Hero!]!","startOffset":0,"duration":0},{"path":["hero","friends",2,"name"],"parentType":"Hero","fieldName":"name","returnType":"String!","startOffset":0,"duration":0},{"path":["hero","friends",1,"name"],"parentType":"Hero","fieldName":"name","returnType":"String!","startOffset":0,"duration":0},{"path":["hero","friends",0,"name"],"parentType":"Hero","fieldName":"name","returnType":"String!","startOffset":0,"duration":0},{"path":["hero","name"],"parentType":"Hero","fieldName":"name","returnType":"String!","startOffset":0,"duration":0}]}}}"""
```

Comment:
Could we add some (artificial) delays in order to see that all the duration/offset are generated correctly as well?

Maybe just change the `Hero` to take `ZIO[Clock, Nothing, String]` and then pass `ZIO.sleep(1.second) *> ZIO.succeed("Han Solo")`

Reply:
Good idea!